opencog/reasoning/RuleEngine/rule-engine-src/pln/BackwardChainer.h

```diff
@@ -50,42 +54,177 @@ using namespace opencog;
+	vector<Type> logical_link_types_ = { AND_LINK, OR_LINK }; //xxx any additional link should be reflected in @method join_premise_vgrounding_maps
+
 	/**
-	 * Initialize target list
+	 * Looks if first list input exists in,VARIABLE_NODE the second ,gets that aren't contained in the second list
+	 * and returns the first amongst the non contained ones
+	 * @param connectors
+	 * @param visited
+	 * @return
 	 */
-public:
-	HandleSeq target; //set of targets
-	HandleSeq rule; //set of matching rules
-	/*
-	 * Initialize the potential target list atom with initial target atom
-	 *Repeat:
-	 * 1.Choose the next,target from the potential-target list (via stochastic selection base on weights that
-	 *  are assigned to potential targets based on multiple factors)
-	 * 2.Choose that rule to apply to the target ( selecting f among the admissbile rules in the rul base via
-	 *  stochastic selection base on the wehths fo the rules in the current context).(Non that each rul asom indicates
-	 *  a fitness function to be used for its premises base non the fitness function for its conclusion)
-	 * 3.Chose premises for the rule via searching the "focus set" specified,suing the pattermatcher(with an appropriate
-	 * callback to guie the PM's  search based on the fitness functions for the premises)
-	 * 4.Check if one fo the stopping criteria has been met.if so exit the loop and stop repeating.
-	 * 5.Push the selected premises an to the  potential target list
-	 * 6.REturn to the start of the loop
+	Handle get_unvisited_logical_link(HandleSeq& connectors,
+			HandleSeq& visited);
+	/**
+	 * matches the variables in the target to their groundings in the variable grounding map list
+	 * @param hgoal the target Handle consisting of variable nodes
+	 * @param var_grounding_map a variable to groundings map list
+	 * @return  a map of variable to all found groundings
+	 */
+	map<Handle, HandleSeq> ground_target_vars(Handle& hgoal,
+			vector<map<Handle, HandleSeq>>& var_grounding_map);
+	/**
+	 * maps a variable to its grounding
+	 * in failing to do so.
+	 * @param htarget the goal
+	 */
+	map<Handle, HandleSeq> do_bc(Handle& htarget);
+	/**
+	 * maps a variable to its grounding
+	 * in failing to do so.
+	 * @param htarget the goal
+	 */
+	map<Handle, HandleSeq> backward_chain( Handle& htarget, Handle& rule);
+	/**
+	 * gets the roolt logical link in an implications link
+	 * eg. (ImplicationLink (AndLink .....) (...)) AndLink will be the root logical link
+	 * @param hrule
+	 * @return a handle if there is a root logical link or Handle::UNDEFINED if there is no
+	 */
+	Handle get_root_logical_link(Handle hrule)
+			throw (opencog::InvalidParamException);
+	/**
+	 * find and return all Handles with no variables
+	 * @param handles
+	 * @return vectors of handles with no variables in them
+	 */
+	HandleSeq get_grounded(HandleSeq handles);
+	/**
+	 * Find and return all handles containing variable and are within an implicationLink
```

Comment:
FYI, there are handy tools for finding variables and nodes and things like that, here:  opencog/query/PatternUtils.h